### DISCUSSION TITLE
Auto layout for nodes with different/custom sizes

### DISCUSSION THREAD

**Discussion by martin-dimi** [xyflow/xyflow#2981] [Q&A] [answered]
Hey all. We've got a setup in which we use d3 to automatically layout our graph as a tree - we followed the pro example on your website to set it up. Currently, this works as all of our nodes have the same hardcoded size, which we can pass into the autoLayout functionality. 

We're looking to introduce new node types which have different sizes (for now we have a set of 3 distinct sizes, but that will probably increase). Do you have any recommendations on how we can approach this? D3 doesn't seem to support it (different sizes). To me this seems like a common usecase and maybe we're missing the obvious solution 🙏

**Comment by bcakmakoglu** [comment]
You might want to check out [d3-flextree.](https://github.com/Klortho/d3-flextree)

**Reply by martin-dimi** [reply]
Oh, that was fast! I think this is exactly what we need! Thanks - will give it a go today

**Reply by martin-dimi** [reply]
@bcakmakoglu any similar libraries come to mind? `d3-flextree` doesn't export types and doesn't seem to be that maintained.

**Reply by bcakmakoglu** [reply, marked as answer]
There is [dagre](https://github.com/dagrejs/dagre) which is marked as deprecated but works very well, supports different sizes and also comes with types in the `@types/dagre` package.

Other than that, maybe [elkjs](https://github.com/kieler/elkjs)? No idea if it supports different node sizes though :/

**Reply by martin-dimi** [reply]
We decided to go with dagre even though it's deprecated. Works really well. Thank you!
